In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [6]:
df = pd.read_csv('mobile_price_range_data.csv',encoding = 'latin1')

In [7]:
df.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

In [8]:
df.shape

(2000, 21)

In [9]:
df.isnull().sum()

battery_power    0
blue             0
clock_speed      0
dual_sim         0
fc               0
four_g           0
int_memory       0
m_dep            0
mobile_wt        0
n_cores          0
pc               0
px_height        0
px_width         0
ram              0
sc_h             0
sc_w             0
talk_time        0
three_g          0
touch_screen     0
wifi             0
price_range      0
dtype: int64

In [10]:
df.duplicated().sum()

0

In [11]:
df.dtypes

battery_power      int64
blue               int64
clock_speed      float64
dual_sim           int64
fc                 int64
four_g             int64
int_memory         int64
m_dep            float64
mobile_wt          int64
n_cores            int64
pc                 int64
px_height          int64
px_width           int64
ram                int64
sc_h               int64
sc_w               int64
talk_time          int64
three_g            int64
touch_screen       int64
wifi               int64
price_range        int64
dtype: object

In [12]:
x = df.drop('price_range',axis = 1)
y = df['price_range']
print(type(x))
print(type(y))
print(x.shape)
print(y.shape)

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>
(2000, 20)
(2000,)


In [13]:
from sklearn.model_selection import train_test_split

In [14]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=30,random_state = 42)
print(x_train.shape)   
print(x_test.shape)   
print(y_train.shape)  
print(y_test.shape)  

(1970, 20)
(30, 20)
(1970,)
(30,)


In [15]:
#log reg
from sklearn.linear_model import LogisticRegression

In [16]:
lr = LogisticRegression()
lr.fit(x_train,y_train)

C:\Users\anish\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [18]:
print('Train score ' , lr.score(x_train,y_train)) 
print('Test sscore', lr.score(x_test,y_test))   

Train score  0.6431472081218275
Test sscore 0.6


In [19]:
ypred_lr = lr.predict(x_test)

In [20]:
print(ypred_lr)

[0 2 0 3 1 2 3 0 3 3 0 1 2 3 3 2 2 2 1 0 0 1 0 2 1 1 3 3 3 0]


In [21]:
from sklearn.metrics import confusion_matrix,classification_report

In [22]:
cm = confusion_matrix(y_test,ypred_lr)
print(cm)
print(classification_report(y_test,ypred_lr))

[[6 2 0 0]
 [2 3 1 1]
 [0 1 4 3]
 [0 0 2 5]]
              precision    recall  f1-score   support

           0       0.75      0.75      0.75         8
           1       0.50      0.43      0.46         7
           2       0.57      0.50      0.53         8
           3       0.56      0.71      0.63         7

    accuracy                           0.60        30
   macro avg       0.59      0.60      0.59        30
weighted avg       0.60      0.60      0.60        30



In [23]:
#KNN classifier
from sklearn.neighbors import KNeighborsClassifier

In [24]:
knn1 = KNeighborsClassifier(n_neighbors = 11)
knn1.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=11)

In [25]:
print('Train score',knn1.score(x_train,y_train))
print('Test score',knn1.score(x_test,y_test))

Train score 0.9512690355329949
Test score 0.9333333333333333


In [26]:
ypred_knn1 = knn1.predict(x_test)
print(ypred_knn1)

[0 2 1 3 1 1 2 0 3 1 0 0 2 3 2 2 3 3 1 0 0 1 1 2 0 1 3 2 2 0]


In [27]:
cm = confusion_matrix(y_test,ypred_knn1)
print(cm)
print(classification_report(y_test,ypred_knn1))

[[8 0 0 0]
 [0 7 0 0]
 [0 1 7 0]
 [0 0 1 6]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         8
           1       0.88      1.00      0.93         7
           2       0.88      0.88      0.88         8
           3       1.00      0.86      0.92         7

    accuracy                           0.93        30
   macro avg       0.94      0.93      0.93        30
weighted avg       0.94      0.93      0.93        30



In [28]:
#svm model - with linear kernal
from sklearn.svm import SVC

In [29]:
svc_m1 = SVC(kernel = 'linear',C=1)
svc_m1.fit(x_train,y_train)

SVC(C=1, kernel='linear')

In [30]:
print('Train Score',svc_m1.score(x_train,y_train))
print('Test Score',svc_m1.score(x_test,y_test))

Train Score 0.9893401015228427
Test Score 0.9333333333333333


In [31]:
ypred_svc_m1 = svc_m1.predict(x_test)
print(ypred_svc_m1)

[0 2 1 3 1 1 2 0 3 1 0 1 2 3 3 2 3 3 1 0 0 2 1 2 0 1 3 3 2 0]


In [32]:
cm = confusion_matrix(y_test,ypred_svc_m1)
print(cm)
print(classification_report(y_test,ypred_svc_m1))

[[7 1 0 0]
 [0 7 0 0]
 [0 0 7 1]
 [0 0 0 7]]
              precision    recall  f1-score   support

           0       1.00      0.88      0.93         8
           1       0.88      1.00      0.93         7
           2       1.00      0.88      0.93         8
           3       0.88      1.00      0.93         7

    accuracy                           0.93        30
   macro avg       0.94      0.94      0.93        30
weighted avg       0.94      0.93      0.93        30



In [33]:
##svm model - with rbf kernal
svc_m1 = SVC(kernel = 'rbf',C=1)
svc_m1.fit(x_train,y_train)

SVC(C=1)

In [34]:
print('Train Score',svc_m1.score(x_train,y_train))
print('Test Score',svc_m1.score(x_test,y_test))

Train Score 0.9548223350253807
Test Score 0.9666666666666667


In [35]:
ypred_svc_m1 = svc_m1.predict(x_test)
print(ypred_svc_m1)

[0 2 1 3 1 1 2 0 3 1 0 0 2 3 3 2 3 3 1 0 0 1 1 2 0 1 3 2 2 0]


In [36]:
print(cm)
print(classification_report(y_test,ypred_svc_m1))

[[7 1 0 0]
 [0 7 0 0]
 [0 0 7 1]
 [0 0 0 7]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         8
           1       0.88      1.00      0.93         7
           2       1.00      0.88      0.93         8
           3       1.00      1.00      1.00         7

    accuracy                           0.97        30
   macro avg       0.97      0.97      0.97        30
weighted avg       0.97      0.97      0.97        30



In [37]:
#Decision Tree Classifier

from sklearn.tree import DecisionTreeClassifier

In [38]:
dt = DecisionTreeClassifier()
dt.fit(x_train,y_train)

DecisionTreeClassifier()

In [39]:
print('Train score ' , dt.score(x_train,y_train))  
print('Test sscore', dt.score(x_test,y_test)) 

Train score  1.0
Test sscore 0.8666666666666667


In [40]:
ypred_dt = dt.predict(x_test)
print(ypred_dt)

[0 2 1 3 1 2 2 0 3 2 0 1 2 3 3 2 3 3 1 0 0 1 1 2 0 1 3 2 2 0]


In [41]:
cm = confusion_matrix(y_test,ypred_dt)
print(cm)
print(classification_report(y_test,ypred_dt))

[[7 1 0 0]
 [0 5 2 0]
 [0 1 7 0]
 [0 0 0 7]]
              precision    recall  f1-score   support

           0       1.00      0.88      0.93         8
           1       0.71      0.71      0.71         7
           2       0.78      0.88      0.82         8
           3       1.00      1.00      1.00         7

    accuracy                           0.87        30
   macro avg       0.87      0.87      0.87        30
weighted avg       0.87      0.87      0.87        30



In [42]:
# Random Forest Classifier
from sklearn.ensemble import RandomForestClassifier

In [43]:
rf = RandomForestClassifier(n_estimators=80,criterion='gini',max_depth = 10,min_samples_split = 15)
rf.fit(x_train,y_train) 

RandomForestClassifier(max_depth=10, min_samples_split=15, n_estimators=80)

In [44]:
print('Train score ' , rf.score(x_train,y_train))  
print('Test sscore', rf.score(x_test,y_test)) 

Train score  0.9771573604060914
Test sscore 0.8333333333333334


In [45]:
ypred_rf = rf.predict(x_test)
print(ypred_rf)

[0 2 1 3 1 2 2 0 3 1 0 1 2 3 2 2 3 3 1 0 0 1 1 2 0 1 3 3 2 0]


In [46]:
cm = confusion_matrix(y_test,ypred_rf)
print(cm)
print(classification_report(y_test,ypred_rf))

[[7 1 0 0]
 [0 6 1 0]
 [0 1 6 1]
 [0 0 1 6]]
              precision    recall  f1-score   support

           0       1.00      0.88      0.93         8
           1       0.75      0.86      0.80         7
           2       0.75      0.75      0.75         8
           3       0.86      0.86      0.86         7

    accuracy                           0.83        30
   macro avg       0.84      0.83      0.84        30
weighted avg       0.84      0.83      0.84        30



<br>REPORT:

   Model Accuracy
   
1)Logistic Regression          0.60  Accuracy

2)KNN classifier               0.93  Accuracy

3)SVM Classifier 
   with linear kernal          0.93  Accuracy
   
4)SVM Classifier with
   rbf kernal                  0.97  Accuracy
   
5)Decision Tree Clasifier      0.87  Accuracy

6)Random Forest Classifier     0.83  Accuracy


     Thus the Model SVM Classifier with RBF Kernal gives the best Accuracy 
     of o.97.
  <br>